bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/api/EnedisHttpApi.java

```diff
@@ -0,0 +1,249 @@
+                throw new LinkyException("Connection failed step 3 - auth1 : " + result.getContentAsString());
+            }
+
+            AuthData authData = gson.fromJson(result.getContentAsString(), AuthData.class);
+            if (!config.username.contentEquals(authData.callbacks.get(0).input.get(0).valueAsString())) {
+                throw new LinkyException("Authentication error, the authentication_cookie is probably wrong");
+            }
+
+            authData.callbacks.get(1).input.get(0).value = config.password;
+            url = "https://mon-compte.enedis.fr/auth/json/authenticate?realm=/enedis&spEntityID=SP-ODW-PROD&goto=/auth/SSOPOST/metaAlias/enedis/providerIDP?ReqID%"
+                    + reqId
+                    + "%26index%3Dnull%26acsURL%3Dhttps://apps.lincs.enedis.fr/saml/SSO%26spEntityID%3DSP-ODW-PROD%26binding%3Durn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST&AMAuthCookie=";
+
+            logger.debug("Step 3 : auth2 - send the auth data");
+            result = httpClient.POST(url).header(HttpHeader.CONTENT_TYPE, "application/json")
+                    .content(new StringContentProvider(gson.toJson(authData))).send();
+            if (result.getStatus() != 200) {
+                throw new LinkyException("Connection failed step 3 - auth2 : " + result.getContentAsString());
+            }
+
+            AuthResult authResult = gson.fromJson(result.getContentAsString(), AuthResult.class);
+            logger.debug("Add the tokenId cookie");
+            addCookie("enedisExt", authResult.tokenId);
+
+            logger.debug("Step 4 : retrieve the SAMLresponse");
+            data = getData(URL_MON_COMPTE + "/" + authResult.successUrl);
+            htmlDocument = Jsoup.parse(data);
+            el = htmlDocument.select("form").first();
+            samlInput = el.select("input[name=SAMLResponse]").first();
+
+            logger.debug("Step 5 : post the SAMLresponse to finish the authentication");
+            result = httpClient.POST(el.attr("action")).content(getFormContent("SAMLResponse", samlInput.attr("value")))
+                    .send();
+            if (result.getStatus() != 302) {
+                throw new LinkyException("Connection failed step 5");
+            }
+            connected = true;
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+            throw new LinkyException("Error opening connection with Enedis webservice", e);
+        }
+    }
+
+    public String getLocation(ContentResponse response) {
+        return response.getHeaders().get(HttpHeader.LOCATION);
+    }
+
+    public void disconnect() throws LinkyException {
+        if (connected) {
+            try { // Three times in a row to get disconnected
+                String location = getLocation(httpClient.GET(URL_APPS_LINCS + "/logout"));
+                location = getLocation(httpClient.GET(location));
+                location = getLocation(httpClient.GET(location));
```

Comment:
Line doubled

Reply:
This is deliberate, Three URLs have to be called to properly log out. A comment is present just above on this.